BetaCameras/Sick.VisionaryT.Pro/VisionaryTPro.cs

```diff
@@ -0,0 +1,340 @@
+                return new ParamDesc<string>()
+                {
+                    Description = "IP address",
+                    ReadableWhen = ParamDesc.ConnectionStates.Disconnected | ParamDesc.ConnectionStates.Connected,
+                    WritableWhen = ParamDesc.ConnectionStates.Disconnected
+                };
+            }
+        }
+
+        public string IPAddress { get; set; }
+
+        public int Port { get; set; } = 2115;
+
+#if !NETSTANDARD2_0
+        public override Icon CameraIcon { get => Properties.Resources.SickIcon; }
+#endif
+
+
+        public VisionaryTPro()
+            : base()
+        {
+            
+        }
+
+        #region MetriCam2 Camera Interface
+
+        public new string Name { get => "Visionary-T-Pro"; }
+
+        protected override void LoadAllAvailableChannels()
+        {
+            ChannelRegistry cr = ChannelRegistry.Instance;
+
+            Channels.Clear();
+
+            Channels.Add(cr.RegisterChannel(ChannelNames.Intensity));
+            Channels.Add(cr.RegisterChannel(ChannelNames.Distance));
+        }
+
+
+        protected override void ConnectImpl()
+        {
+            _updateThreadError = null;
+            _updateThread = new Thread(new ThreadStart(UpdateLoop));
+            _cancelUpdateThreadSource = new CancellationTokenSource();
+
+            if (string.IsNullOrEmpty(IPAddress))
+            {
+                string msg = string.Format("IP address is not set. It must be set before connecting.");
+                log.Error(msg);
+                throw ExceptionBuilder.Build(typeof(ConnectionFailedException), Name, "error_connectionFailed", msg);
+            }
+
+            try
+            {
+                IPAddress camIP = System.Net.IPAddress.Parse(IPAddress);
+                _socket = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp);
+                IPEndPoint ep = new IPEndPoint(camIP, Port);
+                _socket.Connect(ep);
+            }
+            catch(Exception)
+            {
+                string msg = string.Format("Error connecting to IP address.");
+                log.Error(msg);
+                throw ExceptionBuilder.Build(typeof(ConnectionFailedException), Name, "error_connectionFailed", msg);
+            }
+
+            ActivateChannel(ChannelNames.Distance);
+            ActivateChannel(ChannelNames.Intensity);
+            SelectChannel(ChannelNames.Intensity);
+
+            _updateThread.Start();
+        }
+
+        protected override void DisconnectImpl()
+        {
+            _cancelUpdateThreadSource.Cancel();
+            _updateThread.Join();
+            _socket.Close();
+            _frameAvailable.Reset();
+        }
+
+        protected override void UpdateImpl()
+        {
+            _frameAvailable.WaitOne();
+
+            if (null != _updateThreadError)
+            {
+                Disconnect();
+                throw new ImageAcquisitionFailedException(_updateThreadError);
```

Comment:
It would be cool if this would also contain the inner exception that originally caused the `UpdateLoop` to stop.